File: caronte_client/java/Sample.java
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;

import org.caronte.CaronteClient;
import org.json.JSONObject;

public class Sample {
	
	public static String sampleProvider(CaronteClient caronte_client) throws IOException{
		String ticket = caronte_client.getTicket();
		if (ticket != null) {
			URL url = new URL("http://"+caronte_client.getHost()+":"+caronte_client.getPort()+"/provider/");
			HttpURLConnection con = (HttpURLConnection) url.openConnection();
			JSONObject params = new JSONObject();
			params.put("ticket", new JSONObject(ticket));
			con.setRequestMethod("POST");
			con.setDoInput(true);
			con.setDoOutput(true);
			
			// send request
			OutputStream os = con.getOutputStream();
			os.write(params.toString().getBytes("UTF-8"));
			os.close();
			
			// receive response
			InputStream is = con.getInputStream();
			String res = CaronteClient.readAllInput(is);
			JSONObject response = new JSONObject(res);
			is.close();
			String cookie = con.getHeaderField("Set-Cookie");
			con.disconnect();
			if (response.getString("status").equals("OK")){
				String service_provider = caronte_client.setOtherKey(response.getString("key"));
				con = (HttpURLConnection)url.openConnection();
				con.setRequestMethod("GET");
				con.setDoInput(true);
				con.setDoOutput(true);
				con.addRequestProperty("Cookie", cookie);
				con.connect();
				
				// read JSON response
				is = con.getInputStream();
				res = CaronteClient.readAllInput(is);
				response = new JSONObject(res);
				is.close();
				con.disconnect();
				
				// decrypt and return data from service provider
				String data = caronte_client.decryptOther(service_provider, response.getString("msg"));
				return data;
			}
		}
		return null;
	}
	
	public static void main(String[] args){
		
		try {
			// create a connection to local Caronte server
			CaronteClient caronte_client = new CaronteClient("localhost", 8000);
			
			// login using test user
			boolean login_res = caronte_client.login("test@caronte.com", "Caront3Te$t");
			System.out.println("Login: "+login_res);
			if (!login_res) {
				System.out.println("Could not login to server");
				return;
			}
			// print user details
			JSONObject user = caronte_client.getUserDetails();
			if (user != null){
				System.out.println("User name: "+user.getString("name"));
				System.out.println("User email: "+user.getString("email"));
				System.out.println("User joined: "+user.getString("joined"));
			}
			// verify ticket functionality and logout
			System.out.println("Ticket validates: "+caronte_client.validateTicket());
			System.out.println("Invalidate: "+caronte_client.invalidateTicket());
			System.out.println("Validate: "+caronte_client.validateTicket());
			System.out.println("Revalidate: "+caronte_client.revalidateTicket());
			System.out.println("Validate: "+caronte_client.validateTicket());
			System.out.println("Provider Data: "+sampleProvider(caronte_client));
			System.out.println("Logout: "+caronte_client.logout());
		}
		catch (Exception e) {
			System.out.println("Could not connect to Caronte Server");
			e.printStackTrace();
		}
	}

}

File: caronte_client/python/sample.py
from caronte_client import CaronteClient
import json

def loginToProvider(caronte_conn):
	ret = None;
	ticket = caronte_conn.getTicket(); # get a valid ticket
	if (ticket != None):
		# login to service provider using Caronte ticket
		params = json.dumps({"ticket":ticket})
		caronte_conn.conn.request("POST", "/provider/", body=params, headers=caronte_conn.header)
		res = caronte_conn.conn.getresponse()
		if (res.status == 200):
			jres = json.loads(res.read().decode("UTF-8"))
			if (jres["status"] == "OK"):
				# set temporary session key for safe communication
				ret = caronte_conn.setOtherKey(jres["key"])
	return ret

def getProviderData(caronte_conn, provider_id):
	secret_data = None
	caronte_conn.conn.request("GET", "/provider/", headers=caronte_conn.header)
	res = caronte_conn.conn.getresponse()
	if (res.status == 200):
		jres = json.loads(res.read().decode("UTF-8"))
		# decrypt data from service provider
		secret_data = caronte_conn.decryptOther(provider_id, jres["msg"])
	return secret_data

def main():
	caronte_conn = CaronteClient("localhost", 8000)	
	print("Login:", caronte_conn.login("test@caronte.com", "Caront3Te$t"))
	user = caronte_conn.getUserDetails()
	if user != None:
		print("User name:", user["name"])
		print("e-mail:", user["email"])
		print("Joined:", user["joined"])
	print("Ticket validates:", caronte_conn.validateTicket())
	print("Invalidate:", caronte_conn.invalidateTicket())
	print("Validate:", caronte_conn.validateTicket())
	print("Revalidate:", caronte_conn.revalidateTicket())
	print("Validate:", caronte_conn.validateTicket())
	provider_id = loginToProvider(caronte_conn)
	print("Login to provider:", provider_id)
	print("Prover data:", getProviderData(caronte_conn, provider_id))
	print("Logout:", caronte_conn.logout())

if __name__ == "__main__":
	main()
